lib/gc/Transforms/CPUPhysicalRegisterPass.cpp

```diff
@@ -0,0 +1,2978 @@
+
+    for (auto v : operandVector) {
+      if (v == originalResult) {
+        replacedVector.insert(forResult);
+        continue;
+      }
+      replacedVector.insert(v);
+    }
+    getVectorBasedFusion().getGroupOpInitArgs()[startGroup - 1] =
+        replacedVector;
+  }
+}
+
+mlir::FailureOr<scf::ForOp> ForLoopGenerator::generateVectorizedForLoop(
+    const size_t groupId, IRRewriter &rewriter, VectorType vectorType) {
+  // prepare for loop iterargs
+  GenerateLoopHelper loopHelper(groupId, 0);
+  prepareForLoopArgs(groupId, loopHelper);
+
+  ArrayRef<int64_t> shapes = vectorType.getShape();
+  // generate for loop
+  auto forOp = constructNestedForOp(groupId, rewriter, rewriter.getUnknownLoc(),
+                                    shapes, loopHelper);
+  replaceOpUsersWithForLoopResult(forOp, groupId, loopHelper.nextAnchorResults,
+                                  loopHelper.nextAnchorResultsIdxMap,
+                                  loopHelper.nextAnchorResultOrignalResultMap);
+
+  return forOp;
+}
+
+bool LoopGeneratorImpl::isGroupHasSpecialOperation(const size_t grpIdx) {
+  auto &rdCanonicalizer = getMultiRdCanonicalizers()[grpIdx];
+  auto &tpCanonicalizer = getTransposeCanonicalizers()[grpIdx];
+  auto &shapeCastCanonicalizer = getShapeCastCanonicalizers()[grpIdx];
+  return !rdCanonicalizer.getCandidateOps().empty() or
+         !tpCanonicalizer.getCandidateOps().empty() or
+         !shapeCastCanonicalizer.getCandidateOps().empty();
+}
+
+void LoopGeneratorImpl::generateGroupOpVectorizedIR(const int idx) {
+  auto &grp = getVectorBasedFusion().getOpGroups()[idx];
+  if (grp.empty()) {
+    LDBG("Current operation Group is empty.");
+    return;
+  }
+  // TODO: special operation better fusion
+  if (isGroupHasSpecialOperation(idx))
+    return;
+
+  VectorType groupType =
+      getVectorBasedFusion().getGroupBiggestRankVectorType()[idx];
+  IRRewriter rewriter(grp.back());
+  rewriter.setInsertionPointAfter(grp.back());
+  // 1. Rewrite operation as vectorized form
+  // 2. Generate loop
+  // rewriteOperationAsVectorize(rewriter, idx);
+  auto forOp = generateVectorizedForLoop(idx, rewriter, groupType);
+  // special operation do not need to change anything
+  if (failed(forOp))
+    return;
+
+  moveLoopInvariantCode(forOp.value());
+}
+
+/// Pass that lower to physical vector.
+struct CPUPhysicalRegisterPass
+    : public impl::CPUPhysicalRegisterPassBase<CPUPhysicalRegisterPass> {
+
+  void runOnOperation() final {
+    //
+    auto *ctx = &getContext();
+    RewritePatternSet patterns(ctx);
+    auto func = getOperation();
+
+    if (hasNotSupportOperation(&func)) {
+      LDBG("Not support operation appears in current function.");
+      return;
+    }
+    // affineApply operation is always used by other operations.
+    std::function<bool(Operation *)> candidateFunc = isProducerOp;
+    moveSomeInterferenceOperation(&func, ctx, candidateFunc);
+    candidateFunc = isCandidateMoveOperations;
```

Comment:
Can  we consolidate these two functions into single one which are to be moved?
BTW, it's better to rename "moveSomeInterferenceOperation()" to a meaningful one?

Reply:
Sure~ The name of this function will rename as `moveOpsFrontOrBack`. But the first time this method is used is to move some ops that are only used by other ops, such as affine.apply, we need move those operation in the front of the current block. The second time is to move some ops that may interfere analysis. There is another move after current pass result, which is the move of read and write in the IR of the entire function, such as moving the read operation before all the calculation operations to prevent the instruction pipeline from generating bubbles. So this method may need to be called separately for certain ops.